### PR TITLE
React to RoutePattern changes in Routing

### PR DESCRIPTION
**Pull request by rynowak**
This wasn't so hard 😆

### CODE REVIEW

File: test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/MvcEndpointDataSourceTests.cs [hunk 1]
```diff
@@ -71,7 +71,7 @@ public void Endpoints_AccessParameters_InitializedFromProvider()
 
             Assert.Equal(displayName, matcherEndpoint.DisplayName);
             Assert.Equal(order, matcherEndpoint.Order);
-            Assert.Equal(template, matcherEndpoint.Template);
+            Assert.Equal(template, matcherEndpoint.RoutePattern.RawText);
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/MvcEndpointDataSourceTests.cs]: ctrl+H for the win!

File: test/Microsoft.AspNetCore.Mvc.Core.Test/Routing/DispatcherUrlHelperTest.cs [hunk 2]
```diff
@@ -100,14 +99,14 @@ protected override IServiceProvider CreateServices()
...
+            endpoints.Add(CreateEndpoint(null, "home/newaction/{id}", new { id = "defaultid", controller = "home", action = "newaction" }, 1));
+            endpoints.Add(CreateEndpoint(null, "home/contact/{id}", new { id = "defaultid", controller = "home", action = "contact" }, 2));
+            endpoints.Add(CreateEndpoint(null, "home2/newaction/{id}", new { id = "defaultid", controller = "home2", action = "newaction" }, 3));
+            endpoints.Add(CreateEndpoint(null, "home2/contact/{id}", new { id = "defaultid", controller = "home2", action = "contact" }, 4));
+            endpoints.Add(CreateEndpoint(null, "home3/contact/{id}", new { id = "defaultid", controller = "home3", action = "contact" }, 5));
+            endpoints.Add(CreateEndpoint("namedroute", "named/home/newaction/{id}", new { id = "defaultid", controller = "home", action = "newaction" }, 6));
+            endpoints.Add(CreateEndpoint("namedroute", "named/home2/newaction/{id}", new { id = "defaultid", controller = "home2", action = "newaction" }, 7));
+            endpoints.Add(CreateEndpoint("namedroute", "named/home/contact/{id}", new { id = "defaultid", controller = "home", action = "contact" }, 8));
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/Routing/DispatcherUrlHelperTest.cs]: OK so **ALL** of these were invalid. I felt like I'm crazy? None of these really rely on the fact that it's an optional parameter. An optional parameter with a default just behaves like a parameter with a default anyway.

Comment (JamesNK) [reply test/Microsoft.AspNetCore.Mvc.Core.Test/Routing/DispatcherUrlHelperTest.cs]: @kichalla

**Comment by rynowak** [comment]
FYI it's expected right now that this fails. I didn't push the dependencies.props part of these changes to this branch.

File: test/Microsoft.AspNetCore.Mvc.Core.Test/Routing/DispatcherUrlHelperTest.cs (hunk 2 above)
Comment (kichalla) [reply test/Microsoft.AspNetCore.Mvc.Core.Test/Routing/DispatcherUrlHelperTest.cs]: Thanks @JamesNK  Yes, this change makes sense to me

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcEndpointDataSource.cs [hunk 3]
```diff
@@ -298,20 +299,18 @@ public Task RouteAsync(RouteContext context)
...
 
+            var defaults = new RouteValueDictionary(nonInlineDefaults);
+            EnsureRequiredValuesInDefaults(action.RouteValues, defaults);
+
             var metadataCollection = BuildEndpointMetadata(action, routeName, source);
             var endpoint = new MatcherEndpoint(
                 next => invokerDelegate,
-                template,
-                new RouteValueDictionary(nonInlineDefaults),
+                RoutePatternFactory.Parse(template, defaults, constraints: null),
```
Comment (kichalla) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcEndpointDataSource.cs]: Wouldn't this throw if defaults were specified inline as well as non-inline?

Comment (kichalla) [reply src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcEndpointDataSource.cs]: Like would this hit https://github.com/aspnet/Routing/blob/release/2.2/src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternFactory.cs#L215?

Comment (rynowak) [reply src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcEndpointDataSource.cs]: They can't be duplicates of each other - the templates we pass in here never have the 'action' and 'controller' as parameters. This was actually a problem in the benchmark @JamesNK - what it was testing for attribute routing is something that will (and should) throw.

**Pull request merged**